### PR TITLE
Support increasing batch size

### PR DESCRIPTION
**Pull request by markus-hinsche**
Revive the following contribution: https://github.com/tensorflow/tensorflow/pull/45338

**Changes:**
- change the behavior of `data_adapter.DataAdapter` and `data_adapter.DataHandler` to accept batch size that are increasing during training
- add unit test with `keras.utils.Sequence` that increases its batch size from 5 to 10 over the course of 2 epochs
- add unit test which checks correct behavior of `data_adapter.DataAdapter` and `data_adapter.DataHandler`

**Resources:**
- Idea for increasing batch size: https://arxiv.org/abs/1711.00489
- Rasa workaround: https://github.com/RasaHQ/rasa/blob/main/rasa/utils/tensorflow/temp_keras_modules.py#L477
- Rasa BatchSequence which increases batch size over time: https://github.com/RasaHQ/rasa/blob/main/rasa/utils/tensorflow/data_generator.py#L333

### CODE REVIEW

**Review by fchollet — APPROVED** [review]
LGTM, thank you for the contribution.

File: keras/engine/data_adapter.py [hunk 1]
```diff
@@ -958,7 +958,7 @@ def generator_fn():
     return generator_fn
 
   def get_size(self):
-    return self._size
```
[the remark this answers is not in the document]
Comment (markus-hinsche) [reply keras/engine/data_adapter.py:961]: Yes, good catch! I updated the PR accordingly

**Review by fchollet — APPROVED** [review]

**Comment by fchollet** [comment]
There seems to be some test failures on CI: please take a look https://source.cloud.google.com/results/invocations/03af213d-1f66-4d24-9887-4db36eefb21d/targets/keras%2Fgithub%2Fubuntu%2Fcpu%2Fpresubmit/log

**Comment by gbaned** [comment]
Hi @markus-hinsche Can you please check @fchollet's comments and keep us posted ? Thank you!

**Comment by markus-hinsche** [comment]
I will not be able to look into this in the next 12 days. I could look into it after, but I am also happy if somebody wants to help out with this.

**Comment by fchollet** [comment]
No rush at all. I took a quick look at the stack trace, it seems to be that `cardinality` is not defined for `DistributedDataset` -- we should just route around it in our code, and we could also open a feature request for adding support for the above in the future.

**Comment by markus-hinsche** [comment]
I tried to fix the issue following the advice. 
@gbaned Could you trigger a re-run of the tests to we see if they pass now?

**Comment by gbaned** [comment]
> I tried to fix the issue following the advice. @gbaned Could you trigger a re-run of the tests to we see if they pass now?

@markus-hinsche  I have re-ran the tests.  Thank you.

**Comment by markus-hinsche** [comment]
I tried to fix the remaining issue.
@gbaned Could you trigger another re-run of the tests to see if they pass now?

**Comment by markus-hinsche** [comment]
@gbaned @fchollet Seems like the tests are passing now. Let me know if you need anything else from me before merging

**Review by fchollet — APPROVED** [review]
LGTM, thanks!

**Pull request merged**